third_party/move/move-compiler/src/parser/syntax.rs

```diff
@@ -2253,10 +2251,30 @@ fn parse_dot_or_index_chain(context: &mut Context) -> Result<Exp, Box<Diagnostic
             Tok::Period => {
                 context.tokens.advance()?;
                 let n = parse_identifier_or_positional_field(context)?;
-                let ahead = context.tokens.peek();
-                if matches!(ahead, Tok::LParen | Tok::ColonColon) {
-                    let generics = if ahead == Tok::ColonColon {
-                        context.tokens.advance()?;
+                let has_generics = match context.tokens.peek() {
+                    Tok::Less if n.loc.end() as usize == context.tokens.start_loc() => true,
```

Comment:
Note that this is a breaking change.

Both of the code pieces below compiled and ran successfully before this change, but they fail compilation now.

```
module 0xc0ffee::m {
    struct S has copy, drop {
        a: u64
    }

    public fun test(): bool {
        let x = S { a: 42 };
        x.a<9
    }
}
```

```move
module 0xc0ffee::m {
    struct S has copy, drop {
        a: u64
    }

    public fun test(): (bool, bool) {
        let x = S { a: 42 };
        let u9 = 9;
        let u10 = 10;
        (x.a<u9, u10>(23))
    }
}
```

Reply:
Lets discuss in the compiler meeting, I think it is acceptable. I can also make it conditional on 2.2.

Reply:
Another test case to try out if it is handled:

```move
module 0xc0ffee::m {
    enum S1 { Inner }
    enum S2<phantom T> has drop { One }

    fun test1(s: S2<u8>) {
        s is S2<u8>::One;
    }
}
```

It fails parsing on `main`, and we probably should not?

Reply:
Rust uses the following syntax in similar situations: `S2::<u8>::One`.

Reply:
We currently have the postfixing after the variant, here is a testcase: https://github.com/aptos-labs/aptos-core/blob/c54de8eaf5f7acf3ed6f002b6356d9bbca93145a/third_party/move/move-compiler-v2/tests/checking/variants/variants_test_ok.move#L30-L32

Perhaps fine like this.

Right now, a 'name chain' in the parsers AST cannot contain generics, thats why the above wouldn't be easy possible.